docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
 
+```bash
+/var/www/SS_acme > upgrade-code add-namespace "Acme\\Web\\Models" ./mysite/code/Models --write
+/var/www/SS_acme > upgrade-code add-namespace "Acme\\Web\\Tests" ./mysite/tests --write
+/var/www/SS_acme > upgrade-code add-namespace "Acme\\SubModule" ./sub-module/code --write
 ```
-$ ./vendor/bin/sake dev/tasks/MigrateFileTask
+
+### Manually namespacing your codebase
+Go through each PHP file under `mysite/code` and add a `namespace` statement at the top, *with the exception of the files for `Page` or `PageController`*.
+
+Take for example this SilverStripe 3 file located at `mysite/code/Products/ExplosiveTennisBall.php`.
+```php
+<?php
+
+class ExplosiveTennisBall extends DataObject
+{
+    // ...
+}
 ```
 
-## Upgrade tips
+Assuming your root namespace is `Acme\Web`, the equivalent namespaced file will look like this.
+```php
+<?php
+namespace Acme\Web\Products;
 
-While there's some code we can automatically rewrite, other uses of changed SilverStripe APIs aren't that obvious.
-You can use our `inspect` to get some hints on where you need to review code manually.
-Hints will generally point to more detail about a specific upgrade in this guide.
-This task requires [the upgrader tool](https://github.com/silverstripe/silverstripe-upgrader).
+class ExplosiveTennisBall extends DataObject
+{
+    // ...
+}
+```
 
+If you intend to use the upgrader to update references to your namespaced classes, you'll need to create a `mysite/.upgrade.yml` file.
+```yaml
+mappings:
+  ExplosiveTennisBall: Acme\Web\Products\ExplosiveTennisBall
 ```
-~/.composer/vendor/bin/upgrade-code inspect ./mysite
+
+If you intend to manually update references to your namespaced classes, you'll need to go through each of your file to add `use` statements.
+
+For example, if `mysite/code/ProductService.php` is using the `ExplosiveTennisBall` class, you'll need to add a use statement at the top of the file just after it's own namespace definition.
+```php
+<?php
+namespace Acme\Web;
+
+use Acme\Web\Products\ExplosiveTennisBall;
+
+class ProductService
+{
+    // ...
+}
 ```
 
-These hints only cover a part of the upgrade work, but can serve as a good indicator for where to start.
+### Enable PSR-4 auto-loading in your `composer.json` file
+If you have namespaced your project and followed the PSR-4 convention, you have the option to enable PSR-4 auto-loading in your composer.json file.
 
-If you've already had a look over the changelog, you will see that there are some fundamental changes that need to be implemented to upgrade from 3.x. Here's a couple of the most important ones to consider:
+Enabling PSR-4 auto-loading is optional. It will provide better auto-loading of your classes in your development environment and will future proof your code for SilverStripe 5.
```

Comment:
Remove "for SilverStripe 5", we likely won't have time to remove our class loader for that.

Reply:
Done